### PR TITLE
Add Angular

### PR DESCRIPTION
**Pull request by Yberion**
Seems to be kinda slow.
Experimental Zoneless setup (little gain compared with zone.js) to more closely match other frameworks.
(I did not add Hydration as the point is to test SSR only)

```
Running 10s test @ http://localhost:3000
10 connections

┌─────────┬────────┬────────┬────────┬────────┬──────────┬──────────┬────────┐
│ Stat    │ 2.5%   │ 50%    │ 97.5%  │ 99%    │ Avg      │ Stdev    │ Max    │
├─────────┼────────┼────────┼────────┼────────┼──────────┼──────────┼────────┤
│ Latency │ 281 ms │ 302 ms │ 443 ms │ 458 ms │ 319.3 ms │ 45.84 ms │ 465 ms │
└─────────┴────────┴────────┴────────┴────────┴──────────┴──────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev  │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Req/Sec   │ 24      │ 24      │ 31      │ 34      │ 31,1    │ 2,63   │ 24      │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Bytes/Sec │ 4.95 MB │ 4.95 MB │ 6.39 MB │ 7.01 MB │ 6.41 MB │ 541 kB │ 4.95 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

321 requests in 10.12s, 64.1 MB read
```

@JeanMeche @alxhub @alan-agius4 do you think anything could be improved?

### CODE REVIEW

**Comment by garrettld** [comment]
This is consistent with my results on my (quite old and slow) machine, also with zoneless and without hydration:

```
Running 10s test @ http://localhost:3000
10 connections

┌─────────┬────────┬────────┬─────────┬─────────┬───────────┬───────────┬─────────┐
│ Stat    │ 2.5%   │ 50%    │ 97.5%   │ 99%     │ Avg       │ Stdev     │ Max     │
├─────────┼────────┼────────┼─────────┼─────────┼───────────┼───────────┼─────────┤
│ Latency │ 479 ms │ 623 ms │ 1196 ms │ 1239 ms │ 663.01 ms │ 184.58 ms │ 1253 ms │
└─────────┴────────┴────────┴─────────┴─────────┴───────────┴───────────┴─────────┘
┌───────────┬─────┬──────┬─────────┬─────────┬─────────┬─────────┬─────────┐
│ Stat      │ 1%  │ 2.5% │ 50%     │ 97.5%   │ Avg     │ Stdev   │ Min     │
├───────────┼─────┼──────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Req/Sec   │ 0   │ 0    │ 18      │ 20      │ 14.5    │ 6.22    │ 10      │
├───────────┼─────┼──────┼─────────┼─────────┼─────────┼─────────┼─────────┤
│ Bytes/Sec │ 0 B │ 0 B  │ 3.71 MB │ 4.12 MB │ 2.99 MB │ 1.28 MB │ 2.06 MB │
└───────────┴─────┴──────┴─────────┴─────────┴─────────┴─────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

155 requests in 10.03s, 29.9 MB read
```

I'm not very experienced with Angular SSR, but I don't see anything wrong with your implementation. I assume the performance difference is due to DOM emulation.

File: angular/server.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    index: false,
+    wildcard: false,
+  });
+
+  server.get('*', (req, res) => {
+    const { protocol, url, hostname } = req;
+    const baseUrl = protocol + '://' + hostname;
+
+    return commonEngine
+      .render({
```
Comment (alan-agius4) [inline angular/server.ts:29]: If your point to to test SSR only, you may want to disable critical CSS inlining here.

https://angular.dev/guide/ssr#configure-server-side-rendering

Comment (Yberion) [reply angular/server.ts:29]: Thanks for the feedback, this improved a bit the results (see next comment)

**Comment by Yberion** [comment]
**With disabling critical CSS inlining**
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬────────┬────────┬────────┬────────┬───────────┬──────────┬────────┐
│ Stat    │ 2.5%   │ 50%    │ 97.5%  │ 99%    │ Avg       │ Stdev    │ Max    │
├─────────┼────────┼────────┼────────┼────────┼───────────┼──────────┼────────┤
│ Latency │ 202 ms │ 243 ms │ 275 ms │ 434 ms │ 246.96 ms │ 30.62 ms │ 439 ms │
└─────────┴────────┴────────┴────────┴────────┴───────────┴──────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev  │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Req/Sec   │ 35      │ 35      │ 40      │ 46      │ 40,1    │ 2,47   │ 35      │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Bytes/Sec │ 7.21 MB │ 7.21 MB │ 8.25 MB │ 9.49 MB │ 8.26 MB │ 510 kB │ 7.21 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

411 requests in 10.07s, 82.6 MB read
```

**With disabling critical CSS inlining + [`toFixed(2)` + remove `idCounter`](https://github.com/platformatic/ssr-performance-showdown/pull/19)**
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬────────┬────────┬────────┬────────┬───────────┬──────────┬────────┐
│ Stat    │ 2.5%   │ 50%    │ 97.5%  │ 99%    │ Avg       │ Stdev    │ Max    │
├─────────┼────────┼────────┼────────┼────────┼───────────┼──────────┼────────┤
│ Latency │ 197 ms │ 225 ms │ 310 ms │ 332 ms │ 228.04 ms │ 27.53 ms │ 406 ms │
└─────────┴────────┴────────┴────────┴────────┴───────────┴──────────┴────────┘
┌───────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────┬─────────┐
│ Stat      │ 1%      │ 2.5%    │ 50%     │ 97.5%   │ Avg     │ Stdev  │ Min     │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Req/Sec   │ 38      │ 38      │ 42      │ 48      │ 43,4    │ 3,3    │ 38      │
├───────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────┼─────────┤
│ Bytes/Sec │ 5.78 MB │ 5.78 MB │ 6.38 MB │ 7.29 MB │ 6.59 MB │ 501 kB │ 5.77 MB │
└───────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────┴─────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

444 requests in 10.05s, 65.9 MB read
```

**Comment by Yberion** [comment]
Hello @galvez, do you think it's too late to have Angular in the benchmark?

I applied changes from https://github.com/platformatic/ssr-performance-showdown/pull/19, is there anything else to change if it's not too late?

File: angular/angular.json [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            "tsConfig": "tsconfig.app.json",
+            "assets": [
+              {
+                "glob": "**/*",
+                "input": "public"
+              }
+            ],
+            "scripts": [],
+            "server": "src/main.server.ts",
+            "prerender": false,
```
Comment (yutamago) [inline angular/angular.json:33]: With SSG (`prerender: true`) I'm getting much higher throughput.
```suggestion
            "prerender": true,
```

Is this not essentially what the other frameworks also do during SSR?

My results:
```
Running 10s test @ http://localhost:3000
10 connections


┌─────────┬──────┬──────┬───────┬───────┬─────────┬─────────┬───────┐
│ Stat    │ 2.5% │ 50%  │ 97.5% │ 99%   │ Avg     │ Stdev   │ Max   │
├─────────┼──────┼──────┼───────┼───────┼─────────┼─────────┼───────┤
│ Latency │ 7 ms │ 9 ms │ 14 ms │ 15 ms │ 9.58 ms │ 1.74 ms │ 28 ms │
└─────────┴──────┴──────┴───────┴───────┴─────────┴─────────┴───────┘
┌───────────┬────────┬────────┬────────┬────────┬────────┬─────────┬────────┐
│ Stat      │ 1%     │ 2.5%   │ 50%    │ 97.5%  │ Avg    │ Stdev   │ Min    │
├───────────┼────────┼────────┼────────┼────────┼────────┼─────────┼────────┤
│ Req/Sec   │ 931    │ 931    │ 986    │ 1.038  │ 991,2  │ 35,09   │ 931    │
├───────────┼────────┼────────┼────────┼────────┼────────┼─────────┼────────┤
│ Bytes/Sec │ 142 MB │ 142 MB │ 150 MB │ 158 MB │ 151 MB │ 5.32 MB │ 141 MB │
└───────────┴────────┴────────┴────────┴────────┴────────┴─────────┴────────┘

Req/Bytes counts sampled once per second.
# of samples: 10

10k requests in 10.04s, 1.51 GB read
```

File: angular/src/app/app.component.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  template: `
+    <div id="wrapper">
+      @for (tile of this.tiles; track $index) {
+        <div
+          class="tile"
+          [style]="{ left: tile.x.toFixed(2) + 'px', top: tile.y.toFixed(2) + 'px' }"
+        ></div>
+      }
+    </div>
+  `,
```
Comment (yutamago) [inline angular/src/app/app.component.ts:20]: While this doesn't noticeably affect the loading times, the code creates one unnecessary `<div>` too many.

It should be written like this instead:
```suggestion
  host: {
    id: 'wrapper',
  },
  template: `
    @for (tile of this.tiles; track $index) {
      <div
        class="tile"
        [style]="{ left: tile.x.toFixed(2) + 'px', top: tile.y.toFixed(2) + 'px' }"
      ></div>
    }
  `,
```

File: angular/angular.json (hunk 2 above)
Comment (Yberion) [reply angular/angular.json:33]: Hello, this test is about SSR, not SSG

I don't think other framework serves cached or SSG pages on SSR

**Pull request closed without merging**